doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -19,46 +19,195 @@ processes for individual cases where breaking backwards compatibility
+   packages rather than, e.g., appealing to the docs or for stylistic reasons.
+4. Silently getting a wrong answer is much worse than getting a loud error.
+
+When assessing the costs of proposed changes, keep in mind that most users do
+not read the mailing list, do not look at deprecation warnings, and sometimes
+wait more than one or two years before upgrading from their old version. And
+that NumPy has millions of users, so "no one will do or use this" is very
+likely incorrect.
+
+Benefits include improved functionality, usability and performance, as well as
+lower maintenance cost and improved future extensibility.
+
+Fixes for clear bugs are exempt from this backwards compatibility policy.
+However in case of serious impact on users (e.g. a downstream library doesn't
+build anymore or would start giving incorrect results), even bug fixes may have
+to be delayed for one or more releases.
+
+
+Strategies related to deprecations
+----------------------------------
+
+Getting hard data on the impact of a deprecation of often difficult. Strategies
+that can be used to assess such impact include:
+
+- Use a code search engine ([1]_) or static ([2]_) or dynamic ([3]_) code
+  analysis tools to determine where and how the functionality is used.
+- Testing prominent downstream libraries against a development build of NumPy
+  containing the proposed change to get real-world data on its impact.
+- Making a change in master and reverting it, if needed, before a release. We
+  do encourage other packages to test against NumPy's master branch, so this
+  often turns up issues quickly.
+
+If the impact is unclear or significant, it is often good to consider
+alternatives to deprecations. For example discouraging use in documentation
+only, or moving the documentation for the functionality to a less prominent
+place or even removing it completely. Commenting on open issues related to it
+that they are low-prio or labeling them as "wontfix" will also be a signal to
+users, and reduce the maintenance effort needing to be spent.
+
+
+Implementing deprecations and removals
+--------------------------------------
+
+Deprecation warnings are necessary in all cases where functionality
+will eventually be removed.  If there is no intent to remove functionality,
+then it should not be deprecated either. A "please don't use this" in the
+documentation or other type of warning should be used instead.
+
+Deprecations:
+
+- shall include the version number of the release in which the functionality
+  was deprecated.
+- shall include information on alternatives to the deprecated functionality, or a
+  reason for the deprecation if no clear alternative is available.
+- shall use ``VisibleDeprecationWarning`` rather than ``DeprecationWarning``
+  for cases of relevance to end users. For cases only relevant to
+  downstream libraries, a regular ``DeprecationWarning`` is fine.
+  *Rationale: regular deprecation warnings are invisible by default; library
```

Comment:
Is this actually true any more? I think the default visibility changed in 3.8, although possibly not enough.

Reply:
Still true. I think you're referring to [Python Development Mode](https://docs.python.org/3/library/devmode.html#devmode), which makes a number of warnings visible but is not enabled by default.

Reply:
My thinking was that if "end users" have a workflow of running `python myscript.py`, then `DeprecationWarning`s are visible to them automatically.

Reply:
Ah, that's https://www.python.org/dev/peps/pep-0565/ - never even noticed that.

The Python Development Mode docs are then incorrect/incomplete about their discussion of DeprecationWarning.

Reply:
updated this with @seberg's suggestion